### ISSUE TITLE
Socket message poor performance from ESP to a Windows PC

### ISSUE THREAD

**Issue by hmiezin** [me-no-dev/ESPAsyncWebServer#757]
I am using websockets between an ESP8266 and a PC with MS-Windows (sometimes Android or Apple devices)
- When sending data from the client (WEB browser) to the ESP, everything works well, with good performances (~100 bytes messages every 20mS without any loss)
- From ESP to PC, the performance is very poor. I try to send 10 messages / second from ESP to PC. If I check how messages are received by the PC (I use javascript tools), the maximum rate seems to be 1 message every 200-300mS (I did not make accurate measurements but I can do). No error is detected on ESP but the available memory decreases rapidly and the ESP finishes by crashing. If I stop sending messages before the ESP crashes, then the client continues receiving for a while (up to 30sec), and when it stops, no message where lost on the PC (I check frame numbering) and the total memory is recovered on the ESP. This confirm that we have accumulation on the ESP.
Looks like the PC is particularly slow to send acks, and that I do not know how to implement flow control on the ESP (when I send a Ssocket msg, I do not know if it has been sent and acknowldeged)
Last but not least, this is not browser-dependent (luckily !!)

Then, my questions
Does my understanding of the problem seem right ?
Does anyone know any problem with Windows performance when receiving WebSockets ? (the problem is exactly the same with Windows 7 and Windows 10, but the performance is far better on Android)
Does anyone know how to avoid it ? Any Windows update or use of a third party software on the PC ?
If no simple solution can be implemented, I will certainly use events, but size limitation will be an issue.
Thanks in advance for your help.

**Comment by philbowles** [comment]
You are sending too fast. The current version of the lib has a hard-coded limit of #define WS_MAX_QUEUED_MESSAGES 8 on the ESP8266. This *should* stop the esp overloading the server, but it does it by just silently discarding any msg > 8 ...so if the server can't keep up, you will lose data without knowing it.

The running out of memory problem is strange though - the above method was put in (I THINK, I cannot say for sure as I am just a user like you, not the author!) to prevent that when the queue wasn't limited...

In that case, the internal queue would just grow and grow as messages piled up and couldn't be despatched quickly enough...so I'm wondering what version you are using?

Personally I this this approach sucks (and have raised it in other issues, but...) as the only accurate answer to your question is that if you want to send messages that fast, you can't use this library :(

**Comment by hmiezin** [comment]
Hi, philbowles, and thank you for your answer.

I agree with you, I try to send too fast. Or messages are sent too slow. Both means the same.
The question for me was: why 300mS/msg in one direction and less than 20mS/msg in the other ? Especially when we know that websockets keep a persistent connection permanently, then access time should be fast. For me, Webscokets sounds like a dream with "UDP performance and TCP safety". More or less ;o)

The buffer length is not my first concern, because when you fill the bottle faster than you empty it, at one moment, it will overflow. Maybe just the remark that if the buffer for un-sent messages is designed to grow without limits, then any software runs in the wall. I did not see any limit like the one you mentionned (WS_MAX_QUEUED_MESSAGES).

But, I have a question for the SW author. I first want to mention I have a deep respect for the very high quality if the AsyncXXX softwares- and documentation - he gave us those years. Thank you me-no-dev !!
Now the question: as a user, I do not have any access to the success/failure of a sent message, nor transmit buffer size (I did not find anything in the AsyncWebSocket api), then I have zero visibility on the flow control and the possibility buffer explosion. The only way I found to prevent this is to number the frames and monitor the stack, but I do not find this very... satisfying !!

I am not familiar with websockets protocol (I had no time till now to dive in the RFC, but I KNOW I HAVE TO DO IT ;o) 
I beleive that such an asymmetric behaviour is more a question of implementation than a question of protocol.

This delay of 300mS seemed abnormal to me. Could we assume that during those 300mS, the ESP keeps the message in the buffer just because something happened ? Or because something did not happen ?
For instance, if, after sending the frame, ESP expect an acknowledge within 300mS... then the transmitted frame will probably not be discarded from the buffer. 
Is it possible that Windows takes so long before sending and ack ?
Remember, it is not so catastrophic with an Android client !!
All this are just assumptions, I have to investigate with a network analyzer, but have no time at the moment. 

Since last time, just for fun, I tried to answer any websocket message from ESP with a short ack from JAVASCRIPT (I added message date and number to make it reliable)
Then the accepted rate came from 300mS to ~50mS(permanent transmission of short messages from ESP to PC and ACK in return)

Any better understanding/clarification welcome, of course.

Have a nice day !!
